### ISSUE TITLE
Async busy UI

### ISSUE THREAD

**Issue by ermau** [xamarin/Xamarin.PropertyEditing#43]
Need to add the UI restrictions for when async work is going

**Comment by vancura** [comment]
Could you please describe what it's about? I don't know what to design :)

**Comment by ermau** [comment]
It's detailed in the design doc under "Editor States".

**Comment by vancura** [comment]
Let's:

- Hide the control (textfield / checkbox / whatever in the right column), including property button and potential resource icon (in fact everything in the right column), since the user can't edit values there anyway.
- Add the generic spinner (animated).

<img width="408" alt="mac states 2x" src="https://user-images.githubusercontent.com/4982/35802107-212735b0-0a6f-11e8-87cb-96c7d0e06e0e.png">

<img width="408" alt="mac states dark 2x" src="https://user-images.githubusercontent.com/4982/35802110-24c47520-0a6f-11e8-847f-21923ab955f9.png">

Thoughts?

**Comment by CartBlanche** [comment]
I'm personally not a big fan of hiding, then showing things. I would suggest something like leaving the fields there, making them read only (which will also disable the property button ), then maybe changing the highlight colour of the control to yellow (or whatever, but something different to normal to indicate something is happening, or even change the Text to say "loading" ) and use the action button area to show the animating spinner.

**Comment by CartBlanche** [comment]
If we know how long the async operation will take, it would be awesome to have a progress bar there, but I suspect that may be too much to do.

**Comment by vancura** [comment]
I wouldn't do any stroke or highlight, that would be too much (or so-called attention grabber).

What about this?

<img width="441" alt="screen shot 2018-02-05 at 13 03 40" src="https://user-images.githubusercontent.com/4982/35803957-09cb9832-0a76-11e8-9761-3488b326162d.png">

**Comment by alanmcgovern** [comment]
I suspect async gets/sets will typically take a few dozen milliseconds, with a worst case scenario being an unbounded/unknown amount of time.

For the designers/previewer it'll be synchronous as we're reading/writing an in-memory document. The async busy UI will never appear here.

For the Inspector/Workbooks scenario it'll be pretty uncommon to hit the 'async busy' UI as most gets/sets should complete relatively quickly (in comparison to the users typing speed).

I would be worried about the effect hiding the control would have on keyboard focus, similarly for replacing the content of the control with the word 'loading' or equivalent. It could be visually confusing if large visual changes happen for just a few dozen milliseconds and then change back.

What's the simplest thing we can do to support this across all the controls, considering it won't be possible to hit this at all in the main workflows and it'll also be uncommon/rare in the workflow where you can hit it? :)

**Comment by CartBlanche** [comment]
@alanmcgovern I think overlaying a "loading" control might be the easiest to get working, then hide it once the hidden async control is loaded.

@vancura That looks better and clearer, but still think the spinner should be to the right where the notification icons are. That's like the notification area anyway. Seems inconsistent to me to suddenly have it jump to the left. But hey, I'm not a designer, just my tuppence worth
 :)

**Comment by vancura** [comment]
> That looks better and clearer, but still think the spinner should be to the right where the notification icons are.

I tried that, and it was too busy.

<img width="408" alt="mac states 2x" src="https://user-images.githubusercontent.com/4982/35807226-6c8173f0-0a82-11e8-8c6c-886a55f90afd.png">

BW the left-align of the spinner icon is not new, we do have it in the global MS status bar in the toolbar :)

**Comment by vancura** [comment]
One last thing regarding my previous image with right-aligned spinner: there's another problem, in this case, the spinner is too detached from the value which is being loaded. These two should be connected, the label "Loading" and the spinning spinner, as both are two halves of the same progress indication.

**Comment by ermau** [comment]
I realize now my notes on this were too ambiguous:

> We can not accept any value changes from the user while in this state.

This actually means from _any_ property, not just the current one.

**Comment by vancura** [comment]
> This actually means from any property, not just the current one.

That means from ALL properties across the whole Proppy currently displayed?

**Comment by vancura** [comment]
If it's about all properties, we might disable all the content easily.

**Comment by ermau** [comment]
@vancura 

> That means from ALL properties across the whole Proppy currently displayed?

Yes, correct.

**Comment by vancura** [comment]
I'd say in that case just make all the Property Pad contents semitransparent (33%). That should be enough to show it's not editable. Oh and make it not editable.

**Comment by ermau** [comment]
We also want to make sure we cover the cases where:
1. The async delay is taking longer than expected (slow network/issues)
2. Something went boom and it's going to end up timing out.

I'm not sure disabling everything, by itself, buys us enough explanation of what's going on. I'm also not sure flashing the entire UI dim and back every time they change a value is a great experience. Perhaps we only trigger visual changes past a certain time threshold? Maybe add a spinner over it at that time or at a little later delay.

**Comment by vancura** [comment]
> Perhaps we only trigger visual changes past a certain time threshold? 

Yes, that's what I'd do. Take a look at task scheduler in any OS. If the application doesn't respond on time, the application window dims. We can do the same: wait a second (or something) without disabling anything, and when the timeout comes, do this:

<img width="408" alt="mac states 2x" src="https://user-images.githubusercontent.com/4982/35911065-076239d2-0bf9-11e8-8f3d-87c939b6e91a.png">

<img width="408" alt="mac states dark 2x" src="https://user-images.githubusercontent.com/4982/35911076-0a91cf82-0bf9-11e8-94a8-acc47a209ddc.png">

What do you think?
